### PR TITLE
Issue 13

### LINKED ISSUE #13: Implement base type converters
need change way for converting between values. Currently in this library everyfunctions sources use self logic which can be duplicate by any others sources configuration functions. It's no valid way. Need universal converters

### PR DESCRIPTION
**Pull request by kubitre**
now i want to change all methods converts in hocon func for continue works in environment function source. resolve #13

### CODE REVIEW

**Comment by kubitre** [comment]
#17

**Comment by kubitre** [comment]
u can check this

**Comment by kubitre** [comment]
#13

**Comment by artemkaxboy** [comment]
Resolve conflicts, please

**Comment by kubitre** [comment]
> Resolve conflicts, please

check

File: converters/base_types.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return infra.NewGoStructorTrueValue(reflect.ValueOf(float32(source.Int())))
+	case reflect.String:
+		parsed, errParsed := strconv.ParseFloat(source.String(), 32)
+		if errParsed != nil {
+			return infra.NewGoStructorNoValue(destination, errParsed)
+		}
+		return infra.NewGoStructorTrueValue(reflect.ValueOf(float32(parsed)))
+	case reflect.Float32:
+		return infra.NewGoStructorTrueValue(reflect.ValueOf(float32(source.Float())))
```
Comment (artemkaxboy) [inline converters/base_types.go]: Can we use just `return infra.NewGoStructorTrueValue(sourse)` here?

File: converters/kernel_test.go [hunk 2]
```diff
@@ -11,40 +14,932 @@ func TestConvertBetweenPrimitiveTypes(t *testing.T) {
...
+			name: "convert from int into int16 success",
+			args: args{
+				source:      reflect.ValueOf(int(1234)),
+				destination: reflect.ValueOf(int16(0)),
+			},
+			want: infra.NewGoStructorTrueValue(reflect.ValueOf(int16(1234))),
+		},
+		// {
+		// 	name: "convert from",
+		// }
```
Comment (artemkaxboy) [inline converters/kernel_test.go]: Add TODO or throw it away

File: converters/kernel_test.go [hunk 3]
```diff
@@ -11,40 +14,932 @@ func TestConvertBetweenPrimitiveTypes(t *testing.T) {
...
+			args: args{
+				source:      reflect.ValueOf(struct{ field string }{field: "test"}),
+				destination: reflect.ValueOf(int16(0)),
+			},
+			want: infra.NewGoStructorNoValue(reflect.ValueOf(int16(0)),
+				errors.New("can not converted from this type: "+reflect.Struct.String()+" beacuse this type not supported")),
+		},
+		// {
+		// 	name: "convert from",
+		// }
```
Comment (artemkaxboy) [inline converters/kernel_test.go]: Add TODO or throw it away

Comment (artemkaxboy) [reply converters/kernel_test.go]: Here and further

File: infra/cf_value.go [hunk 4]
```diff
@@ -26,15 +26,15 @@ func NewNotAValue(field interface{}, err error) *NotAValue {
 	}
 }
 
-/*NewGoStructorTrueValue - generate new GoStructorValue with completed preparing field
+/*infra.NewGoStructorTrueValue - generate new GoStructorValue with completed preparing field
```
Comment (artemkaxboy) [inline infra/cf_value.go]: You don't need prefix here

File: infra/cf_value.go [hunk 5]
```diff
@@ -26,15 +26,15 @@ func NewNotAValue(field interface{}, err error) *NotAValue {
...
  */
 func NewGoStructorTrueValue(value reflect.Value) GoStructorValue {
 	return GoStructorValue{
 		Value: value,
 	}
 }
 
-/*NewGoStructorNoValue - generate new GoStructorValue with error handling value
+/*infra.NewGoStructorNoValue - generate new GoStructorValue with error handling value
```
Comment (artemkaxboy) [inline infra/cf_value.go]: You don't need prefix here

File: converters/base_types.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+		}
+		return infra.NewGoStructorTrueValue(reflect.ValueOf(int(convertedValue)))
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		fmt.Println("Level: Debug. Message: start convert value ", source.String(), " into int type")
+		if source.Type().ConvertibleTo(destination.Type()) {
+			return infra.NewGoStructorTrueValue(source.Convert(destination.Type()))
+		}
+		return infra.NewGoStructorNoValue(destination, errors.New("can not be converted from "+source.Kind().String()+" into int"))
+	default:
+		return infra.NewGoStructorNoValue(destination, errors.New("can not converted from this type: "+source.Kind().String()+" beacuse this type not supported"))
```
[the remark this answers is not in the document]
Comment (kubitre) [reply converters/base_types.go]: ++

File: converters/base_types.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+	}
+}
+
+func convertToInt8(source reflect.Value, destination reflect.Value) infra.GoStructorValue {
+	switch source.Kind() {
+	case reflect.String:
+		fmt.Println("Level: Debug. Message: start convert value ", source.String(), " into int8 type")
+		convertedValue, errorConvert := strconv.ParseInt(source.String(), 10, 8)
+		if errorConvert != nil {
+			return infra.NewGoStructorNoValue(destination, errors.New("can not converted to this type: "+destination.Kind().String()))
```
[the remark this answers is not in the document]
Comment (kubitre) [reply converters/base_types.go]: ++

File: converters/base_types.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+		}
+		return infra.NewGoStructorTrueValue(reflect.ValueOf(int8(convertedValue)))
+	case reflect.Int, reflect.Int16, reflect.Int32, reflect.Int64:
+		fmt.Println("Level: Debug. Message: start convert value ", source.String(), " into int8 type")
+		if source.Type().ConvertibleTo(destination.Type()) {
+			return infra.NewGoStructorTrueValue(source.Convert(destination.Type()))
+		}
+		return infra.NewGoStructorNoValue(destination, errors.New("can not be converted from "+source.Kind().String()+" into int8"))
+	default:
+		return infra.NewGoStructorNoValue(destination, errors.New("can not converted from this type: "+source.Kind().String()+" beacuse this type not supported"))
```
[the remark this answers is not in the document]
Comment (kubitre) [reply converters/base_types.go]: ++

File: converters/base_types.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+	}
+}
+
+func convertToInt16(source reflect.Value, destination reflect.Value) infra.GoStructorValue {
+	switch source.Kind() {
+	case reflect.String:
+		fmt.Println("Level: Debug. Message: start convert value ", source.String(), " into int16 type")
+		convertedValue, errorConvert := strconv.ParseInt(source.String(), 10, 16)
+		if errorConvert != nil {
+			return infra.NewGoStructorNoValue(destination, errors.New("can not converted to this type: "+destination.Kind().String()))
```
[the remark this answers is not in the document]
Comment (kubitre) [reply converters/base_types.go]: ++

File: converters/kernel_test.go (hunk 2 above)
Comment (kubitre) [reply converters/kernel_test.go]: ++

File: converters/kernel_test.go (hunk 3 above)
Comment (kubitre) [reply converters/kernel_test.go]: ++

File: infra/cf_value.go (hunk 4 above)
Comment (kubitre) [reply infra/cf_value.go]: ++

File: infra/cf_value.go (hunk 5 above)
Comment (kubitre) [reply infra/cf_value.go]: ++

File: converters/base_types.go (hunk 1 above)
Comment (kubitre) [reply converters/base_types.go]: +=

**Pull request merged**